File: scripts/resync-extension-frontends.mjs
#!/usr/bin/env node
/**
 * Re-install extensions from file_registry to copy /ext/ frontend bundles.
 * Grants backend Commit on the frontend canister when needed.
 *
 * Usage:
 *   node scripts/resync-extension-frontends.mjs <backend> [--frontend <id>] [--registry <id>]
 */
import fs from 'node:fs';
import { HttpAgent, Actor } from '@dfinity/agent';
import { IDL } from '@dfinity/candid';
import { Principal } from '@dfinity/principal';
import { Secp256k1KeyIdentity } from '@dfinity/identity-secp256k1';

const DEFAULT_REGISTRY = 'iebdk-kqaaa-aaaau-agoxq-cai';
const DEFAULT_PEM = '/root/.config/dfx/identity/deployer/identity.pem';

const assetIdl = ({ IDL: idl }) =>
  idl.Service({
    grant_permission: idl.Func(
      [
        idl.Record({
          to_principal: idl.Principal,
          permission: idl.Variant({
            Commit: idl.Null,
            Prepare: idl.Null,
            ManagePermissions: idl.Null,
          }),
        }),
      ],
      [],
      [],
    ),
  });

const realmIdl = ({ IDL: idl }) =>
  idl.Service({
    list_runtime_extensions: idl.Func([], [idl.Text], ['query']),
    install_extension_from_registry: idl.Func([idl.Text], [idl.Text], []),
  });

function parseArgs(argv) {
  const opts = {
    registry: DEFAULT_REGISTRY,
    frontend: null,
    pem: DEFAULT_PEM,
    ext: null,
  };
  const positional = [];
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--registry') opts.registry = argv[++i];
    else if (a === '--frontend') opts.frontend = argv[++i];
    else if (a === '--identity-pem') opts.pem = argv[++i];
    else if (a === '--ext') opts.ext = argv[++i];
    else positional.push(a);
  }
  opts.backend = positional[0];
  if (!opts.backend) throw new Error('backend canister id required');
  return opts;
}

async function main() {
  const opts = parseArgs(process.argv);
  const identity = Secp256k1KeyIdentity.fromPem(fs.readFileSync(opts.pem, 'utf8'));
  console.log('Caller:', identity.getPrincipal().toText());
  console.log('Backend:', opts.backend);

  const agent = new HttpAgent({ host: 'https://icp0.io', identity });
  await agent.fetchRootKey();

  const realm = Actor.createActor(realmIdl, {
    agent,
    canisterId: Principal.fromText(opts.backend),
  });

  const listed = JSON.parse(await realm.list_runtime_extensions());
  const extIds = opts.ext
    ? [opts.ext]
    : listed.runtime_extensions || listed.installed || [];
  const manifests = listed.all_manifests || {};

  let frontend = opts.frontend;
  if (!frontend) {
    const canisters = listed?.canisters || [];
    frontend = canisters.find((c) => c.canister_type === 'frontend')?.canister_id;
  }
  if (!frontend) {
    throw new Error('frontend canister id not found — pass --frontend');
  }
  console.log('Frontend:', frontend);
  console.log('Registry:', opts.registry);

  const asset = Actor.createActor(assetIdl, {
    agent,
    canisterId: Principal.fromText(frontend),
  });
  console.log(`\nGranting Commit on frontend to backend...`);
  await asset.grant_permission({
    to_principal: Principal.fromText(opts.backend),
    permission: { Commit: null },
  });
  console.log('✓ Commit granted\n');

  console.log(`Resyncing ${extIds.length} extension(s)...`);
  let ok = 0;
  let fail = 0;
  for (const extId of extIds) {
    const version = manifests[extId]?.version ?? '?';
    process.stdout.write(`→ ${extId}@${version} ... `);
    const payload = JSON.stringify({
      registry_canister_id: opts.registry,
      ext_id: extId,
      version: null,
      frontend_canister_id: frontend,
    });
    try {
      const raw = await realm.install_extension_from_registry(payload);
      const result = JSON.parse(raw);
      if (result.success) {
        console.log(`✓ (${result.frontend_files_copied ?? '?'} files)`);
        ok++;
      } else {
        console.log(`✗ ${result.error}`);
        fail++;
      }
    } catch (e) {
      console.log(`✗ ${e.message ?? e}`);
      fail++;
    }
  }
  console.log(`\nDone: ${ok} ok, ${fail} failed`);
  if (fail) process.exit(1);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/resync_extension_frontends.py
#!/usr/bin/env python3
"""Re-install extensions from file_registry to restore /ext/ frontend assets."""

from __future__ import annotations

import argparse
import json
import sys
import time

from ic.agent import Agent
from ic.candid import Types, encode
from ic.client import Client
from ic.identity import Identity

DEFAULT_REGISTRY = "iebdk-kqaaa-aaaau-agoxq-cai"
DEFAULT_FRONTEND = "fcm3z-5qaaa-aaaac-bfq4a-cai"
DEFAULT_PEM = "/root/.config/dfx/identity/deployer/identity.pem"
DEFAULT_HOST = "https://icp0.io"


def _text_arg(value: str) -> bytes:
    return encode([{"type": Types.Text, "value": value}])


def _decode_text(raw) -> str:
    if isinstance(raw, list) and raw:
        return raw[0].get("value", "")
    if isinstance(raw, str):
        return raw
    raise ValueError(f"Unexpected candid decode result: {raw!r}")


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("backend", help="Realm backend canister id")
    parser.add_argument("--registry", default=DEFAULT_REGISTRY)
    parser.add_argument("--frontend", default=DEFAULT_FRONTEND)
    parser.add_argument("--identity-pem", default=DEFAULT_PEM)
    parser.add_argument("--ext", help="Single extension id (default: all installed)")
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    ident = Identity.from_pem(open(args.identity_pem, encoding="utf-8").read())
    agent = Agent(ident, Client(DEFAULT_HOST))
    print(f"Caller:   {ident.sender()}")
    print(f"Backend:  {args.backend}")
    print(f"Registry: {args.registry}")
    print(f"Frontend: {args.frontend}")

    listed_raw = agent.query_raw(
        args.backend,
        "list_runtime_extensions",
        encode([]),
        [{"type": Types.Text}],
    )
    listed = json.loads(_decode_text(listed_raw))
    # list_runtime_extensions uses runtime_extensions; older payloads used installed.
    installed = listed.get("runtime_extensions") or listed.get("installed") or []
    manifests = listed.get("all_manifests") or {}
    ext_ids = [args.ext] if args.ext else installed

    if not ext_ids:
        print("No extensions installed.")
        return 0

    print(f"Resyncing {len(ext_ids)} extension(s)...")
    ok = fail = 0

    for ext_id in ext_ids:
        version = (manifests.get(ext_id) or {}).get("version", "?")
        print(f"\n→ {ext_id}@{version}")
        if args.dry_run:
            continue

        payload = json.dumps(
            {
                "registry_canister_id": args.registry,
                "ext_id": ext_id,
                "version": None,
                "frontend_canister_id": args.frontend,
            }
        )
        try:
            result_raw = agent.update_raw(
                args.backend,
                "install_extension_from_registry",
                _text_arg(payload),
                [{"type": Types.Text}],
                timeout=300,
            )
            result = json.loads(_decode_text(result_raw))
            if result.get("success"):
                copied = result.get("frontend_files_copied", "?")
                print(f"  ✓ synced v{result.get('version', version)} ({copied} frontend files)")
                ok += 1
            else:
                print(f"  ✗ {result.get('error', result)}")
                fail += 1
        except Exception as exc:
            print(f"  ✗ {exc}")
            fail += 1
        time.sleep(1)

    print(f"\nDone: {ok} ok, {fail} failed")
    return 1 if fail else 0


if __name__ == "__main__":
    sys.exit(main())
